### ISSUE TITLE
execution problem in Python 3.5.6 "'str' object has no attribute 'decode'" 

### ISSUE THREAD

**Issue by alonsorm77** [daducci/AMICO#86]
Hello AMICO guys,

I am getting this execution error when calling `generate_kernels(self, regenerate, lmax)` (Python 3.5.6 in Ubuntu 18.04.2 LTS):

`~/anaconda3/envs/dipy35/lib/python3.5/site-packages/amico/lut.py in <dictcomp>(.0)`
`     75      if sys.version_info.major == 3:`
`     76               aux = pickle.load( rotation_matrices_file, fix_imports=True, encoding='bytes' )`
`---> 77              result_aux = {k.decode(): v for k, v in aux.items()}`
`     78             return result_aux`
`AttributeError: 'str' object has no attribute 'decode'`

It seems that Python 3 is dealing with the strings differently than previous version. 

It could be avoided by changing `k.decode()` to `k` on line 77 of **lut.py**, but i do not know if that is the most proper solution.

Kindly regards.

**Comment by fullbat** [comment]
Hi @alonsorm77,
I've performed all the tutorial, (so even the `generate_kernels()`  command)  using python 2.7.15 and python 3.6.8 on Ubuntu 18.04 LTS, importing AMICO installed with the corresponding pip2 and pip3.
I've checked which method was used for handling the string in the two cases and it performed as expected using the `.decode()` during the execution of python 3 and return directly the `pickle.load(rotation_matrices_file)` on the other case.
![AMICO_py](https://user-images.githubusercontent.com/33122065/59452760-706ca280-8ddc-11e9-8fa9-137847b3a537.png)

**Comment by daducci** [comment]
Thanks @fullbat for looking into this! Actually, here there is a discussion about this big change in _interpreting strings_ between Py2 and Py3. Besides, we need to understand (as you had pointed out to me) that the new code takes much more time to execute in Py3!!

**Comment by fullbat** [comment]
@daducci the difference between python 2 and 3 is that `str()`  in python 2 has ASCII  types, separate `unicode()`, but no byte type, while python 3 has Unicode (utf-8) strings, and 2 byte classes: byte and bytearrays.

From `pickle` documentation:

>  Currently the default protocol is 3; a backward-incompatible protocol designed for Python 3.0.

The default pickling protocol used by default in Python 3 is incompatible with the protocol used by Python 2.

**Comment by daducci** [comment]
So, does the problem arise if someone creates some ‘pickled’ data with Py2 and then opens it with Py3?

**Comment by fullbat** [comment]
Yes, Python 3 pickling is not able to handle the previous version. You should write the pickled data with a lower protocol number in Python 3. Python 3 introduced a new protocol with the number 3 (and uses it as default), so switch back to a value of 2 which can be read by Python 2.

**Comment by ianhinder** [comment]
The code in AMICO has already been adapted to write the pickle files in "protocol 2" for Python 2 compatibility:
```python
    with open( filename, 'wb+' ) as fid :
        pickle.dump( AUX, fid, protocol=2 )
```
I think that this means that str objects should be written out as 8-bit strings, just as Python 2 would, but the documentation is very vague.  When reading these back in, the "encoding='bytes'" in 
```python
        if sys.version_info.major == 3:
            aux = pickle.load( rotation_matrices_file, fix_imports=True, encoding='bytes' )
```
should convert these into bytes objects, and the decode in
```python
           result_aux = {k.decode(): v for k, v in aux.items()}
```
should decode the bytes objects into standard python 3 unicode str objects, which is what they were originally.  This was added in PR #81 in response to issue #74.

I have tested, however, and find that the files written by python 2 and python 3 with protocol=2 are different.  If you pickle a str object with python 2 and read it back with python 3 and encoding='bytes', you get a bytes object.  This is probably the situation the author of #74 was in, with an existing precomputed matrices file.  *HOWEVER*: if you pickle a str object with python 3 and read it back with encoding='bytes', you end up with a str object again!  

```bash
$ python2 -c 'import pickle; pickle.dump("foo", open("written-by-python2.pickle","wb"), protocol=2)'
$ python3 -c 'import pickle; pickle.dump("foo", open("written-by-python3.pickle","wb"), protocol=2)'

$ python3 -c 'import pickle; print(pickle.load(open("written-by-python2.pickle","rb"), encoding="bytes"))'
b'foo'
$ python3 -c 'import pickle; print(pickle.load(open("written-by-python3.pickle","rb"), encoding="bytes"))'
foo
```
So I think the correct fix here is to be permissive about what is found in the pickle file, and only decode bytes objects.  I have created PR #90 with my proposed fix.
